### ISSUE TITLE
PDF/A conversion/validation

### ISSUE THREAD

**Issue by hhrutter** [hhrutter/pdfcpu#34]
As requested by @tcurdt

It would be great if pdfcpu could be used to convert to PDF/A

https://en.wikipedia.org/wiki/PDF/A

**Comment by magegu** [comment]
hi there,

i'm interested in this feature, too - and I would like to add "merging" to the PFA/A "wishlist" tracked in this issue.

As a first test I merged two valid PFA/A files (pdfcpu: v0.3.4 dev) and run [this validator]( https://www.pdf-online.com/osa/validate.aspx) on the result. The feedback was: 

```
Validating file "out.pdf" for conformance level pdfa-2u
The separator before 'endstream' must be an EOL.
The separator before 'endstream' must be an EOL.
The separator before 'endstream' must be an EOL.
The separator before 'endstream' must be an EOL.
The document does not conform to the requested standard.
The file format (header, trailer, objects, xref, streams) is corrupted.
The document does not conform to the PDF/A-2u standard.
```

I would like to contribute but i don't know a lot about the pdf spec to add valuable info at this point. Is there anything specific you can point me to where to start to debug these issues @hhrutter?

**Comment by hhrutter** [comment]
As soon as pdfcpu has support for PDF/A merging should work out of the box.
Basically you have to make a pass over the cross reference table after reading the file where you check/enforce compliance with PDF/A for all elements which may fail or succeed for any given PDF input file. You would need the PDF/A spec for that. I don't know if that is for free because I have not had the time to look into this because I was busy otherwise.

**Comment by OrganicChem** [comment]
Hi there, any update on this issue?

I'm trying to run pdfcpu within a NodeJS environment,  and if a file is invalid, I would like some guidance on how to convert the PDF to a valid one instream.

**Comment by OrganicChem** [comment]
Hello Horst, here are 4 bad files I am working with:
[bad1.pdf](https://github.com/pdfcpu/pdfcpu/files/5096653/bad1.pdf)
[bad2.pdf](https://github.com/pdfcpu/pdfcpu/files/5096654/bad2.pdf)
[bad3.pdf](https://github.com/pdfcpu/pdfcpu/files/5096655/bad3.pdf)
[bad4.pdf](https://github.com/pdfcpu/pdfcpu/files/5096656/bad4.pdf)

I issue: pdfcpu optimize bad1.pdf and results shown in terminal.

Thanks!

**Comment by hhrutter** [comment]
Thanks for sharing, I'll check them out!

**Comment by hhrutter** [comment]
I had to delete your problematic files because some of them include personal information + they are not regarding this issue.
Rest assured I am analysing them but please refrain from posting files with sensible information here.
It's safer to submit such files to hhrutter@gmail.cm

Thank you!

**Comment by OrganicChem** [comment]
I've managed to get 3 out of the 4 bad files (file 2 remains troublesome) to merge using the following workaround:

Since I'm working mainly in a NodeJS environment, I basically split the problematic files using Hummus, a NodeJS module for Creating, Parsing and Modifying PDF files and streams. After splitting, 3 of the 4 bad files were merged successfully with pdfcpu.

**Comment by hhrutter** [comment]
Sounds good!

In any case I already have fixes for all four of your bad files.
Will be part of the upcoming release.

**Comment by OrganicChem** [comment]
Any timeline on the next release?

**Comment by hhrutter** [comment]
This weekend.

**Comment by OrganicChem** [comment]
Thank you so much Horst for resolving those bad files during a merge process with the new version 0.3.5!

This is much appreciated, I really mean that!

**Comment by hhrutter** [comment]
Once pdfcpu goes beta this will be tackled.
Please keep checking #235.

**Comment by fank** [comment]
## Current State of PDF/A Support - Detailed Analysis

Hi there! I've conducted a comprehensive analysis of the current PDF/A support in pdfcpu (v0.11.0 dev) and wanted to share my findings with the community.

### TL;DR: Infrastructure Exists, PDF/A-Specific Logic Missing

pdfcpu has the foundational components needed for PDF/A support, but currently lacks the PDF/A-specific validation and creation capabilities.

### What Currently EXISTS (Infrastructure):

✅ **OutputIntent Dictionary Validation** (`pkg/pdfcpu/validate/xReftable.go:499-578`)
- Validates structure of OutputIntent dictionaries
- However, does NOT validate PDF/A-specific subtypes (`GTS_PDFA1`, `GTS_PDFA2`, `GTS_PDFA3`, `GTS_PDFA4`)

✅ **XMP Metadata Parsing** (`pkg/pdfcpu/model/metadata.go`)
- XML parsing implemented for Dublin Core and XMP properties
- However, missing PDF/A identification metadata (`pdfaid:part`, `pdfaid:conformance`, `pdfaid:rev`)

⚠️ **ICC Profile Parsing** (`pkg/pdfcpu/iccProfile.go`)
- Basic ICC profile parsing exists
- **CRITICAL NOTE from code (line 27-29)**: "ICC profiles are not yet supported! We fall back to the alternate color space..."
- This is essential for PDF/A-1 compliance

✅ **Font, Color Space, Structure Tree Validation**
- All the necessary validation infrastructure exists
- Just needs PDF/A-specific enforcement rules

### What DOES NOT EXIST:

❌ No PDF/A subtype detection (cannot identify if a PDF claims to be PDF/A)
❌ No conformance level validation (Level A, B, U, E, F)
❌ No PDF/A-specific validation rules:
  - Encryption forbidden in PDF/A-1 (currently allowed)
  - Transparency forbidden in PDF/A-1 (currently allowed)
  - Mandatory font embedding (not enforced)
  - Device color spaces without OutputIntent (not checked)
❌ No PDF/A creation/conversion API
❌ No CLI commands for PDF/A operations

### ZUGFeRD Use Case - Why I Need This

I'm working with **ZUGFeRD** (German e-invoicing standard), which **requires PDF/A-3 compliance**. ZUGFeRD embeds structured XML invoice data into PDF/A-3 documents, and full PDF/A-3 support is mandatory for:

- Embedding XML attachments with `/AFRelationship` set to "Alternative"
- Ensuring PDF/A-3b/3u conformance (visual appearance + Unicode text)
- Meeting EN 16931 requirements for EU e-invoicing interoperability
- Digital signature support (PAdES)

**Without PDF/A-3 support, pdfcpu cannot be used for ZUGFeRD invoice generation or validation.**

### I'd Like to Contribute

I'm very interested in working on this feature, particularly because:
1. I need it for ZUGFeRD compliance
2. The infrastructure is already in place - it's "just" connecting the dots

**Would the maintainers be open to contributions for PDF/A support?** I'd be happy to start with Phase 1 (detection) and work incrementally through the phases.

I've documented my full analysis in a detailed report if anyone is interested in the technical specifics.

Looking forward to feedback from @hhrutter and the community! 💚

**Comment by hhrutter** [comment]
PDF/A validation is not on the roadmap for getting to Beta.

**Comment by tcurdt** [comment]
> PDF/A validation is not on the roadmap for getting to Beta.

@hhrutter can you clarify validation vs creation - neither? or could it become part of the roadmap with contributions?

**Comment by hhrutter** [comment]
That's both the same as intended with this issue.
This will be the very first thing that gets tackled once we are in Beta.
For Beta missing is signature creation which is being worked on right now.
